src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,305 @@
+    B2 = single_qubit_decomposition[4]
+
+    # C (with default target 0) is at index 6
+    C = single_qubit_decomposition[6]
+
+    # last is the global phase gate which is changed to phasegate
+    phase_angle = single_qubit_decomposition[-1].arg_value
+    phase_gate = Gate(
+        "PHASEGATE",
+        targets=[1],
+        arg_value=phase_angle,
+        arg_label=r"{:0.2f} \times \pi".format(phase_angle / np.pi),
+    )
+
+    # change default targets for gates A, B, C
+    change_targets = [A1, A2, X_gate, B1, B2, C]
+    for i in range(len(change_targets)):
+        change_targets[i].targets = [0]
+
+    # form full gate list
+    final_gate_list = [
+        A1,
+        A2,
+        X_gate,
+        CNOT_gate,
+        X_gate,
+        B1,
+        B2,
+        X_gate,
+        CNOT_gate,
+        X_gate,
+        C,
+        phase_gate]
+    return(final_gate_list)
+
+
+def _find_index_for_firstq_target(two_level_array_output, num_qubits):
+    """ Find the two-level array index with first qubit as target and rest
+    as controls.
+    """
+    len_two_level = len(two_level_array_output)
+    single_step = []
+
+    # create an array of single step in gray code sequence
+    for i in range(len_two_level):
+        index1, index2 = two_level_array_output[i][0]
+        gray_code_info = _gray_code_gate_info(index1, index2, num_qubits)
+        if len(gray_code_info) == 1:
+            single_step.append(i)
+
+    # find the index with target_qubit = 0
+    for i in range(len(single_step)):
+        ind = single_step[i]
+        index1, index2 = two_level_array_output[ind][0]
+        gray_code = _gray_code_gate_info(index1, index2, num_qubits)
+        gray_code_target = gray_code[0]['targets =']
+        if gray_code_target[0] == 0:
+            return(ind)
+
+
+
+def decompose_general_qubit_gate(input_gate, num_qubits):
+    check_gate(input_gate, num_qubits)
+    two_level_array = _decompose_to_two_level_arrays(
+        input_gate, num_qubits, expand=False)
+
+    # first object in two level array is a two-level gate with last qubit as
+    # target and others as control
+    last_target_qubit_array = two_level_array[0]
+    # because arrays are returned in a reversed order
+    last_target_qubit_array_index = last_target_qubit_array[0]
+    if last_target_qubit_array_index != [2**num_qubits-2, 2**num_qubits-1]:
+        raise MethodError("The chosen method cannot be used.")
+
+    # look for the two-level matrix with first qubit as target and rest as
+    # controls i.e. want target qubit = 0
+    ind_for_0_target = _find_index_for_firstq_target(
+        two_level_array, num_qubits)
+    first_target_qubit_array = two_level_array[ind_for_0_target]
+    first_target_qubit_array_index = first_target_qubit_array[0]
+
+    # create a list of all two-level arrays without first or last qubit as
+    # targets
+    index_list = list(range(len(two_level_array)))
+    index_list.remove(0)
+    index_list.remove(ind_for_0_target)
```

Comment:
For the other two-level arrays - SWAP gates were used to map to either last qubit as target or first qubit as target (whichever has least number of gates) and then SWAP back to the original two-level array.

Reply:
Re-defining the function